{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computer Simulation of Self Fertilization & Natural Selection\n",
    " ## By Haim Elbaz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    " ## Problem Presentation\n",
    " \n",
    " In class we briefly discussed the negative consqeunces of inbreeding in regard to lethal mutations and alleles. Self fertilization ('selfing') is essentially the most extreme form of inbreeding.\n",
    " Intuitively, it seems that self fertilization cannot increase the mean fitness of a population. However, self fertilization is relatively abundant in plants, especially in crop & model plants. In these species, some part of the population reproduces by self fertilization as opposed to out-crossing. \n",
    " \n",
    " How is that the case? Why was self-fertilization not lost during the evolution of these species? \n",
    " Can it help favoring a genotype that otherwise would not be in favor?\n",
    " \n",
    " In other words - what is the relationship between natural selection and self fertilization? "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ## Model\n",
    "The model can be seen as constructed from two parts. The first part is the effect of self fertilization on genotype frequencies, and the second is the consideration of fitness in the self-fertilization population.\n",
    "\n",
    "Allele frequencies are not changed by self fertilization, as seen in the following equation, so they are not being taken into account. This is replicable by the random genetic drift version of the model at the end. \n",
    "\n",
    "$$ p_1 = P_1 + \\frac{1}{2} H_1 = (P_0 + \\frac{1}{4} H_0) + \\frac{1}{4} H_0 = p_0 $$\n",
    "\n",
    "\n",
    "<u>Genotype Frequencies </u>\n",
    "\n",
    "Assuming Hardy-Weinberg proportions, the following equations represent the change in genotype frequency in regard to self-fertilization, where <i>S</i> is the proportion of self-fertilization in the population and <i>T</i> is the complementary proportion of out-crossing. \n",
    "\n",
    "The alleles and their corresponding frequencies are <i>p</i> & <i>q</i> (<i>p</i> being the dominant allele) and <i>P</i>,<i>Q</i> are their corresponding homozygotes - <i>H</i> being the heterozygote genotype.\n",
    "\n",
    "\n",
    "$$ P_1 = Tp^2_0 + S(P_0 + \\frac{1}{4} H_0) :: (1.a) $$\n",
    "$$ Q_1 = Tq^2_0 + S(Q_0 + \\frac{1}{4} H_0) :: (1.b) $$\n",
    "$$ H_1 = 2Tp_0 q_0 + \\frac{1}{2}S H_0      :: (1.c) $$\n",
    "\n",
    "\n",
    "<u>Natural Selection </u>\n",
    "\n",
    "First, the notation being used for natural selection is relative fitness based, using dominance degree and selection coefficient <i>h</i>,<i>s</i> respectively, where <i>P</i> is the relatively fitter allele.\n",
    "\n",
    "$$ \\overline{w}_{P} = 1    :: (2.a) $$ \n",
    "$$ \\overline{w}_{H} = 1-hs :: (2.b) $$ \n",
    "$$ \\overline{w}_{Q} = 1-s  :: (2.c) $$ \n",
    "\n",
    "Since self-fertilization is a form of inbreeding, it is possible to use the equations for genotype fitness values under inbreeding. Since the inbreeding parameter, <i>f</i>, represends the proportion of inbreeding in the population, and the self-fertilization parameter, <i>S</i>, represents the proportion of self-fertilization, assuming that self-fertilization exists in the population but it is the only form of inbreeding that exists for that population, then we can represent that assumption as  <i>S = f</i>.\n",
    "\n",
    "Accordingly, the mean fitness under no self-fertilization and under self-fertilization in degree <i>f</i> are: \n",
    "\n",
    "$$ \\overline{w} = p^2 w_P + 2pqw_H + q^2 w_Q                :: (3.a)$$ \n",
    "$$ \\overline{w}_{f} = \\overline{w} + fpq(w_P -2 w_H + w_Q ) :: (3.b)$$ \n",
    "\n",
    "Combining equation sets 2 & 3, the extent in which inbreeding/selfing will reduce the mean fitness - termed as the inbreeding depression, the difference between them is:\n",
    "\n",
    "$$ \\overline{w} - \\overline{w}_{f} = sfpq(1-2h):: (4)  $$ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Libraries\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import random"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The Code\n",
    "This version of the model is deterministic and shows the average result over infinite repeats, or the result of an infinite population. The is also a second version that is driven by random genetic drift and the end of the code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ========================================================================== #\n",
    "# Deterministic function for self fertilization & natural selection\n",
    "# ========================================================================== #\n",
    "\n",
    "def det_self_fertilization_simulation(A1_frequency,A2_dominance_degree,\n",
    "                                      selection_coefficient,self_fert_factor,\n",
    "                                      no_of_generations):\n",
    "    # Pointers\n",
    "    p           = A1_frequency\n",
    "    h           = A2_dominance_degree\n",
    "    s           = selection_coefficient\n",
    "    S           = self_fert_factor\n",
    "    generations = range(no_of_generations)\n",
    "    \n",
    "    # Arithmetic parameters\n",
    "    q = 1-p\n",
    "    f = S         # Assuming all inbreeding is self fertilization \n",
    "    T = 1-S\n",
    "    \n",
    "    # Hardy-Weinberg genotype frequencies\n",
    "    P0 = p**2\n",
    "    Q0 = q**2\n",
    "    H0 = 2*p*q\n",
    "    \n",
    "    #Creating lists for parameters + H-W check\n",
    "    p_list, q_list, P1_list, Q1_list, H1_list, hwcheck = [],[],[],[],[],[]\n",
    "    p_list.append(p)\n",
    "    q_list.append(q)\n",
    "    P1_list.append(P0)\n",
    "    Q1_list.append(Q0)\n",
    "    H1_list.append(H0) \n",
    "    # Natural selection parameters with regard to self fertilization\n",
    "    w_11     = 1.0\n",
    "    w_12     = 1-(h*s)\n",
    "    w_22     = 1-s\n",
    "    w_bar    = ((p**2*w_11)+(2*p*q*w_12)+(q**2*w_22))\n",
    "    w_bar_f  = w_bar + f*p*q*(w_11+w_22-2*w_12)\n",
    "    \n",
    "    # Accounting for self fertilization \n",
    "    P1 = ((T*p**2+S*(P0+(0.25*H0)))*w_11)/w_bar_f\n",
    "    Q1 = ((T*q**2+S*(Q0+(0.25*H0)))*w_22)/w_bar_f\n",
    "    H1 = ((2*T*p*q+0.5*S*H0)*w_12)/w_bar_f\n",
    "    P1_list.append(P1)\n",
    "    Q1_list.append(Q1)\n",
    "    H1_list.append(H1) \n",
    "\n",
    "    \n",
    "    # Loop over generations\n",
    "    for generation in generations:      \n",
    "        # Creating new genotype frequencies and saving them, \n",
    "        # this time they depend on their previous value.\n",
    "        P1 = ((T*p**2+S*(P1+(0.25*H1)))*w_11)/w_bar_f\n",
    "        Q1 = ((T*q**2+S*(Q1+(0.25*H1)))*w_22)/w_bar_f\n",
    "        H1 = ((2*T*p*q+0.5*S*H1)*w_12)/w_bar_f\n",
    "        checksum = P1+H1+Q1\n",
    "        P1 = P1/checksum\n",
    "        Q1 = Q1/checksum\n",
    "        H1 = H1/checksum\n",
    "        p  = P1 + 0.5*H1\n",
    "        q  = 1-p\n",
    "        p_list.append(p)\n",
    "        q_list.append(q)\n",
    "        P1_list.append(P1)\n",
    "        Q1_list.append(Q1)\n",
    "        H1_list.append(H1)\n",
    "        hwcheck.append(P1+H1+Q1)\n",
    "        \n",
    "    return P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results\n",
    "\n",
    "\n",
    "## Effect of self-fertilization coefficient on genotype frequencies\n",
    "\n",
    "<u>Total Self-Fertilization</u>\n",
    "\n",
    "First, lets test the model for the effect of self-fertilization with only and with no natural selection. For this purpuse, we will set the selection coefficient to 0.0 and neutralize the effect of outcrossing by setting the selfing factor to 1.0 (varying p for demonstration).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Selfing only, varying p\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.1,\n",
    "        A2_dominance_degree   = 0.0,\n",
    "        selection_coefficient = 0.0,\n",
    "        self_fert_factor      = 1.0,\n",
    "        no_of_generations     = 10)\n",
    "\n",
    "# Plot\n",
    "fig, ax = plt.subplots(2,2, figsize=(10,7), constrained_layout=True)\n",
    "ax[0,0].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[0,0].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[0,0].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[0,0].set_title('Genotype Frequency, p = 0.1' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[0,0].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,0].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,0].set_ylim([0,1.0])\n",
    "ax[0,0].set_xlim([0,generations[-1]])\n",
    "ax[0,0].grid(axis='y', linestyle= '--')\n",
    "ax[0,0].grid(axis='x', linestyle= '--')\n",
    "ax[0,0].set_xticks(generations[::1])\n",
    "ax[0,0].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 0.0,\n",
    "        selection_coefficient = 0.0,\n",
    "        self_fert_factor      = 1.0,\n",
    "        no_of_generations     = 10)\n",
    "\n",
    "# Plot\n",
    "ax[0,1].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[0,1].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[0,1].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[0,1].set_title('Genotype Frequency, p = 0.3' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[0,1].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,1].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,1].set_ylim([0,1.0])\n",
    "ax[0,1].set_xlim([0,generations[-1]])\n",
    "ax[0,1].grid(axis='y', linestyle= '--')\n",
    "ax[0,1].grid(axis='x', linestyle= '--')\n",
    "ax[0,1].set_xticks(generations[::1])\n",
    "ax[0,1].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.6,\n",
    "        A2_dominance_degree   = 0.0,\n",
    "        selection_coefficient = 0.0,\n",
    "        self_fert_factor      = 1.0,\n",
    "        no_of_generations     = 10)\n",
    "\n",
    "# Plot\n",
    "ax[1,0].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[1,0].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[1,0].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[1,0].set_title('Genotype Frequency, p = 0.6' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[1,0].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,0].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,0].set_ylim([0,1.0])\n",
    "ax[1,0].set_xlim([0,generations[-1]])\n",
    "ax[1,0].grid(axis='y', linestyle= '--')\n",
    "ax[1,0].grid(axis='x', linestyle= '--')\n",
    "ax[1,0].set_xticks(generations[::1])\n",
    "ax[1,0].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.9,\n",
    "        A2_dominance_degree   = 0.0,\n",
    "        selection_coefficient = 0.0,\n",
    "        self_fert_factor      = 1.0,\n",
    "        no_of_generations     = 10)\n",
    "\n",
    "\n",
    "# Plot\n",
    "ax[1,1].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[1,1].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[1,1].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[1,1].set_title('Genotype Frequency, p = 0.9' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[1,1].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,1].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,1].set_ylim([0,1.0])\n",
    "ax[1,1].set_xlim([0,generations[-1]])\n",
    "ax[1,1].grid(axis='y', linestyle= '--')\n",
    "ax[1,1].grid(axis='x', linestyle= '--')\n",
    "ax[1,1].set_xticks(generations[::1])\n",
    "ax[1,1].legend()\n",
    "\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that the proportions of the homozygotes P&Q have risen to the exact extent in which the heterozygote declined. Eventually, they both reach equilibrium where the homozygote frequency exactly equals its respective allele frequency and the heterozygote frequency (asymptotically) reaces 0 (P=p, Q=q, H=0). This decay of heterozygosity happens because on average, in each generation the progeny for a heterozygote offspring from heterozygote selfing is only 0.5 of the previous generation, and the other half of the progeny are homozygotes - 0.25 for each homozygote. Equation for said decay:\n",
    "\n",
    "$$ H_t = (\\frac{1}{2})^t H_0 $$\n",
    "\n",
    "\n",
    "<u>Partial Self Fertilization</u>\n",
    "\n",
    "\n",
    "With any degree of partial self-fertilization (S<1.0, testing S = 0.9):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Recreating figure 8.11 from 'Genetics of Populations' by P.W. Hendrick, p. 460\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 0.0,\n",
    "        selection_coefficient = 0.0,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 8)\n",
    "# Plot\n",
    "fig, ax = plt.subplots()\n",
    "plt.plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "plt.plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "plt.plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "plt.title('Genotype Frequency' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "plt.xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "plt.ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax.set_ylim([0,0.7])\n",
    "ax.set_xlim([0,generations[-1]])\n",
    "ax.grid(axis='y', linestyle= '--')\n",
    "ax.grid(axis='x', linestyle= '--')\n",
    "ax.set_xticks(generations)\n",
    "ax.legend()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Heterozygosity is lowered but does not reach zero in equilibrium because some degree of outcrossing (T) exist in the population, in addition to heterozygote population from heterozygote selfing, there is also a heterozygote population resulting from crossing the two homozygotes.\n",
    "\n",
    "## Effect of selection & dominance on a self-fertilizing population's genotype frequencies\n",
    "\n",
    "Lets consider the effect of selection and dominance degrees in continuation to the previous example.\n",
    "\n",
    "<u>Selection</u>\n",
    "\n",
    "\n",
    "Comparing the results for various degrees of selfing where P is slightly advantageous over Q (s = 0.2) and the heterozygote H is exactly in the middle (h = 0.5):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Varying S\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 0.5,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.0,\n",
    "        no_of_generations     = 40)\n",
    "\n",
    "# Plot\n",
    "fig, ax = plt.subplots(2,2, figsize=(10,7), constrained_layout=True)\n",
    "ax[0,0].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[0,0].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[0,0].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[0,0].set_title('Genotype Frequency, S = 0.0' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[0,0].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,0].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,0].set_ylim([0,1.0])\n",
    "ax[0,0].set_xlim([0,generations[-1]])\n",
    "ax[0,0].grid(axis='y', linestyle= '--')\n",
    "ax[0,0].grid(axis='x', linestyle= '--')\n",
    "ax[0,0].set_xticks(generations[::10])\n",
    "ax[0,0].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 0.5,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.333,\n",
    "        no_of_generations     = 40)\n",
    "\n",
    "# Plot\n",
    "ax[0,1].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[0,1].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[0,1].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[0,1].set_title('Genotype Frequency, S = 1/3' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[0,1].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,1].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,1].set_ylim([0,1.0])\n",
    "ax[0,1].set_xlim([0,generations[-1]])\n",
    "ax[0,1].grid(axis='y', linestyle= '--')\n",
    "ax[0,1].grid(axis='x', linestyle= '--')\n",
    "ax[0,1].set_xticks(generations[::10])\n",
    "ax[0,1].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 0.5,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.666,\n",
    "        no_of_generations     = 40)\n",
    "\n",
    "# Plot\n",
    "ax[1,0].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[1,0].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[1,0].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[1,0].set_title('Genotype Frequency, S = 2/3' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[1,0].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,0].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,0].set_ylim([0,1.0])\n",
    "ax[1,0].set_xlim([0,generations[-1]])\n",
    "ax[1,0].grid(axis='y', linestyle= '--')\n",
    "ax[1,0].grid(axis='x', linestyle= '--')\n",
    "ax[1,0].set_xticks(generations[::10])\n",
    "ax[1,0].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 0.5,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 1.0,\n",
    "        no_of_generations     = 40)\n",
    "\n",
    "\n",
    "# Plot\n",
    "ax[1,1].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[1,1].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[1,1].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[1,1].set_title('Genotype Frequency, S = 1.0' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[1,1].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,1].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,1].set_ylim([0,1.0])\n",
    "ax[1,1].set_xlim([0,generations[-1]])\n",
    "ax[1,1].grid(axis='y', linestyle= '--')\n",
    "ax[1,1].grid(axis='x', linestyle= '--')\n",
    "ax[1,1].set_xticks(generations[::10])\n",
    "ax[1,1].legend()\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since the q allele is harmful, it is expected that p will fixate, even that it started as the lower frequency allele and even without selfing. However, we can see that the introduction of selfing into this system have caused fixation to happen earlier (in terms of generations), and also caused the heterozygote population to decline much faster. The more prevalent self-fertilization is, the faster the population fixes and heterozygosity declines.\n",
    "\n",
    "<u>Partial & Complete Dominance</u>\n",
    "\n",
    "After seeing the effect of self-fertilization on the heterozygote genotype, it does not seem we can expect any dramatic change in varying degrees of dominance regardless of selection, the heterozygote is expected to asymptotically reach zero. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Partial & complete dominance, 0<=h<=1\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 0.0,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 10)\n",
    "\n",
    "# Plot\n",
    "fig, ax = plt.subplots(2,2, figsize=(10,7), constrained_layout=True)\n",
    "ax[0,0].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[0,0].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[0,0].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[0,0].set_title('Genotype Frequency, h = 0.0' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[0,0].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,0].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,0].set_ylim([0,1.0])\n",
    "ax[0,0].set_xlim([0,generations[-1]])\n",
    "ax[0,0].grid(axis='y', linestyle= '--')\n",
    "ax[0,0].grid(axis='x', linestyle= '--')\n",
    "ax[0,0].set_xticks(generations[::1])\n",
    "ax[0,0].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 0.333,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 10)\n",
    "\n",
    "# Plot\n",
    "ax[0,1].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[0,1].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[0,1].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[0,1].set_title('Genotype Frequency, h = 1/3' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[0,1].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,1].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,1].set_ylim([0,1.0])\n",
    "ax[0,1].set_xlim([0,generations[-1]])\n",
    "ax[0,1].grid(axis='y', linestyle= '--')\n",
    "ax[0,1].grid(axis='x', linestyle= '--')\n",
    "ax[0,1].set_xticks(generations[::1])\n",
    "ax[0,1].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 0.666,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 10)\n",
    "\n",
    "# Plot\n",
    "ax[1,0].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[1,0].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[1,0].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[1,0].set_title('Genotype Frequency, h = 2/3' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[1,0].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,0].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,0].set_ylim([0,1.0])\n",
    "ax[1,0].set_xlim([0,generations[-1]])\n",
    "ax[1,0].grid(axis='y', linestyle= '--')\n",
    "ax[1,0].grid(axis='x', linestyle= '--')\n",
    "ax[1,0].set_xticks(generations[::1])\n",
    "ax[1,0].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 1.0,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 10)\n",
    "\n",
    "\n",
    "# Plot\n",
    "ax[1,1].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[1,1].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[1,1].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[1,1].set_title('Genotype Frequency, h = 1.0' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[1,1].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,1].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,1].set_ylim([0,1.0])\n",
    "ax[1,1].set_xlim([0,generations[-1]])\n",
    "ax[1,1].grid(axis='y', linestyle= '--')\n",
    "ax[1,1].grid(axis='x', linestyle= '--')\n",
    "ax[1,1].set_xticks(generations[::1])\n",
    "ax[1,1].legend()\n",
    "\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "as h gets closer to 1, the H genotype decays faster, but the effect is on a very small scale. In all of the above examples the population reaches fixation of P, the small case is for the effect of varying degrees of partial dominance to be noticable. \n",
    "\n",
    "This is the case for partial dominance.\n",
    "\n",
    "However, it is noticable that although H is more fit than Q, due to the high degree of self-fertilization it does not become more prevalent. As we have seen in the previous example, it is not always the case and for lower degree of selfing, H did have a higher frequency than Q, although temporarily. \n",
    "\n",
    "<u>Underdominance</u>\n",
    "\n",
    "H is the least fit, in this case the heterozygote is being lost faster:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Underdominance h>0\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 0.0,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 30)\n",
    "\n",
    "# Plot\n",
    "fig, ax = plt.subplots(2,2, figsize=(10,7), constrained_layout=True)\n",
    "ax[0,0].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[0,0].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[0,0].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[0,0].set_title('Genotype Frequency, h = 0.0' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[0,0].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,0].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,0].set_ylim([0,1.0])\n",
    "ax[0,0].set_xlim([0,generations[-1]])\n",
    "ax[0,0].grid(axis='y', linestyle= '--')\n",
    "ax[0,0].grid(axis='x', linestyle= '--')\n",
    "ax[0,0].set_xticks(generations[::10])\n",
    "ax[0,0].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 1.0,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 30)\n",
    "\n",
    "# Plot\n",
    "ax[0,1].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[0,1].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[0,1].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[0,1].set_title('Genotype Frequency, h = 1.0' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[0,1].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,1].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,1].set_ylim([0,1.0])\n",
    "ax[0,1].set_xlim([0,generations[-1]])\n",
    "ax[0,1].grid(axis='y', linestyle= '--')\n",
    "ax[0,1].grid(axis='x', linestyle= '--')\n",
    "ax[0,1].set_xticks(generations[::10])\n",
    "ax[0,1].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 2.0,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 30)\n",
    "\n",
    "# Plot\n",
    "ax[1,0].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[1,0].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[1,0].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[1,0].set_title('Genotype Frequency, h = 2.0' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[1,0].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,0].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,0].set_ylim([0,1.0])\n",
    "ax[1,0].set_xlim([0,generations[-1]])\n",
    "ax[1,0].grid(axis='y', linestyle= '--')\n",
    "ax[1,0].grid(axis='x', linestyle= '--')\n",
    "ax[1,0].set_xticks(generations[::10])\n",
    "ax[1,0].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 3.0,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 30)\n",
    "\n",
    "\n",
    "# Plot\n",
    "ax[1,1].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[1,1].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[1,1].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[1,1].set_title('Genotype Frequency, h = 3.0' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[1,1].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,1].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,1].set_ylim([0,1.0])\n",
    "ax[1,1].set_xlim([0,generations[-1]])\n",
    "ax[1,1].grid(axis='y', linestyle= '--')\n",
    "ax[1,1].grid(axis='x', linestyle= '--')\n",
    "ax[1,1].set_xticks(generations[::10])\n",
    "ax[1,1].legend()\n",
    "\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<u>Overdominance</u>\n",
    "\n",
    "H the fittest so it is lost more slowly, but still lost due to self fertilization:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Overdominance h<0\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = 0.0,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 30)\n",
    "\n",
    "# Plot\n",
    "fig, ax = plt.subplots(2,2, figsize=(10,7), constrained_layout=True)\n",
    "ax[0,0].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[0,0].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[0,0].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[0,0].set_title('Genotype Frequency, h = 0.0' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[0,0].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,0].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,0].set_ylim([0,1.0])\n",
    "ax[0,0].set_xlim([0,generations[-1]])\n",
    "ax[0,0].grid(axis='y', linestyle= '--')\n",
    "ax[0,0].grid(axis='x', linestyle= '--')\n",
    "ax[0,0].set_xticks(generations[::10])\n",
    "ax[0,0].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = -0.5,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 30)\n",
    "\n",
    "# Plot\n",
    "ax[0,1].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[0,1].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[0,1].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[0,1].set_title('Genotype Frequency, h = -0.5' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[0,1].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,1].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,1].set_ylim([0,1.0])\n",
    "ax[0,1].set_xlim([0,generations[-1]])\n",
    "ax[0,1].grid(axis='y', linestyle= '--')\n",
    "ax[0,1].grid(axis='x', linestyle= '--')\n",
    "ax[0,1].set_xticks(generations[::10])\n",
    "ax[0,1].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = -1.0,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 50)\n",
    "\n",
    "# Plot\n",
    "ax[1,0].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[1,0].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[1,0].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[1,0].set_title('Genotype Frequency, h = -1.0' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[1,0].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,0].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,0].set_ylim([0,1.0])\n",
    "ax[1,0].set_xlim([0,generations[-1]])\n",
    "ax[1,0].grid(axis='y', linestyle= '--')\n",
    "ax[1,0].grid(axis='x', linestyle= '--')\n",
    "ax[1,0].set_xticks(generations[::10])\n",
    "ax[1,0].legend()\n",
    "\n",
    "P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = det_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.3,\n",
    "        A2_dominance_degree   = -2.0,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.9,\n",
    "        no_of_generations     = 50)\n",
    "\n",
    "\n",
    "# Plot\n",
    "ax[1,1].plot(generations, P1_list[:-2], label = 'P homozygote', color = 'blue')\n",
    "ax[1,1].plot(generations, Q1_list[:-2], label = 'Q homozygote', color = 'red')\n",
    "ax[1,1].plot(generations, H1_list[:-2], label = 'H Heterozygote', color = 'green')\n",
    "ax[1,1].set_title('Genotype Frequency, h = -2.0' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[1,1].set_xlabel('No. of generations' , fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,1].set_ylabel('Genotype Frequency', fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,1].set_ylim([0,1.0])\n",
    "ax[1,1].set_xlim([0,generations[-1]])\n",
    "ax[1,1].grid(axis='y', linestyle= '--')\n",
    "ax[1,1].grid(axis='x', linestyle= '--')\n",
    "ax[1,1].set_xticks(generations[::10])\n",
    "ax[1,1].legend()\n",
    "\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Considering equations 3.b and 4:\n",
    "\n",
    "$$ \\overline{w}_{f} = \\overline{w} + fpq(w_P -2 w_H + w_Q ) :: (3.b)$$ \n",
    "$$ \\overline{w} - \\overline{w}_{f} = sfpq(1-2h):: (4)  $$ \n",
    "\n",
    "The parameters s,f,p,q are all relative and are always between 0.0 and 1.0. however, h can be greater than 1.0 and also negative. This possibly means that in cases of underdominance where h is negative, selfing could increase the average fitness of the population.\n",
    "\n",
    "Relating the fitness values to each other shows: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting w_bar and w_bar_f for h \n",
    "\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "# Parameters\n",
    "p = 0.3\n",
    "q = 1 - p\n",
    "s = 0.2\n",
    "f = 0.9\n",
    "\n",
    "# Create lists for plot\n",
    "w_bar_list   = []\n",
    "w_bar_f_list = []\n",
    "hh = np.linspace(-3.0, 3.0, 10000)\n",
    "\n",
    "for h in hh:\n",
    "    w_11     = 1.0\n",
    "    w_12     = 1-(h*s)\n",
    "    w_22     = 1-s\n",
    "    w_bar   = ((p**2*w_11)+(2*p*q*w_12)+(q**2*w_22))\n",
    "    w_bar_f = w_bar + f*p*q*(w_11+w_22-2*w_12)\n",
    "    w_bar_list.append(w_bar)\n",
    "    w_bar_f_list.append(w_bar_f)\n",
    "\n",
    "# Plot:\n",
    "fig, ax = plt.subplots()\n",
    "plt.plot(hh, w_bar_list,label = 'w_bar', color = 'blue')\n",
    "plt.plot(hh, w_bar_f_list, label = 'w_bar_f', color = 'red')\n",
    "plt.title('Mean Fitness against Q dominance degree' , fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "plt.xlabel('h')\n",
    "plt.ylabel('Population Fitness')\n",
    "ax.set_ylim([0,1.5])\n",
    "ax.set_xlim([-3,3])\n",
    "ax.set_axisbelow(True)\n",
    "ax.legend()\n",
    "textbox = 'p = '+str(p)+'; s = '+str(s)+'; f = '+str(f)\n",
    "ax.text(0.1, 0.1, textbox)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "According to the above plot, it seems that 'inbreeding depression' could turn over, and when the heterozygote is the least fit allele, a population with self fertilization is more fit on average than a population with no self fertilization. \n",
    "\n",
    "\n",
    "## Summary\n",
    "\n",
    "One of my original ideas was to check if there was a case in which selfing will help a population in which a less fit allele is the more abundant one. \n",
    "\n",
    "According to the model, it seems that self fertilization by itself cannot determine which allele will fix. Without selection, the homozygotes will gradually reach their respective allele frequencies. When selection is introduced, the more fit allele is always the one destined to be fixed ,and in the presense of selfing heterozygosity will always decay. However, selfing could haste fixation & heterozygote decay, which could be dramatically helpful if the genetic drift is very strong (for example, when the population is very small) or if the heterozygote is less fit than both alleles.\n",
    "\n",
    "The combined effect of both partial self-fertilization and natural selection in a population is strongly dependent on the state of the homozygote and the dominance relations between the alleles. In cases of partial-dominance and under-dominance, fixation of the better allele happens faster, causing an increase in the average fitness of the whole population in comparison to a population without self fertilization, which will have to suffer longer from having a larger amount of less-fit individuals for a longer period of time. In cases of over-dominance, the opposite, 'intuitive' case is true - a population without selfing is more fit.  \n",
    "\n",
    "\n",
    "I tried implementing genetic drift into the model, but ended up leaving the model with no randomization, giving a consistent result each time, making it easier to discuss and less easy to 'cherry pick' a favorite result. The downside is that population size effect combined with selfing & selection cannot be directly tested with the model as presented above, but can be tested in the 'random' version of the model, presented below. This was done mainly because during testing of many different parameter sets I did not notice any surprising effect of population size on allele/phenotype frequency except the expected faster and slower fixations in smaller and larger populations, respectively. Most of the repeats, in any population size, produces a very similar result to the above 'deterministic' version. Repeats that vary from it are increasing with a smaller population size.\n",
    "\n",
    "There are several questions that are still open, but can be answered with minor tweaks to the core model. It would be interesting to check what would happen if we correlate the degree of selfing to another parameter. For example, we can set the tested locus to be a gene responsible for selfing itself, where the degree of dominance of the selfing allele is the proportion of selfing in the population (i.e correlate h and S), or make selfing of a certain genotype dependent on its frequency (correlate p/q to S). Also, the model has only 2 alleles for simplicity (there are always 2 parents) but many plants that self pollinate have polyploidity, and might have more than 2 alleles of one locus in their genome. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (Sidenote) Random-driven, genetic drift Version of the model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ========================================================================== #\n",
    "# Random function for self fertilization & natural selection\n",
    "# ========================================================================== #\n",
    "\n",
    "def rand_self_fertilization_simulation(A1_frequency,A2_dominance_degree,\n",
    "                                  selection_coefficient,self_fert_factor,\n",
    "                                  population_size,no_of_generations):\n",
    "\n",
    "    # Pointers\n",
    "    p           = A1_frequency\n",
    "    h           = A2_dominance_degree\n",
    "    s           = selection_coefficient\n",
    "    S           = self_fert_factor\n",
    "    N           = population_size\n",
    "    generations = range(no_of_generations)\n",
    "    \n",
    "    # Arithmetic parameters\n",
    "    q = 1-p\n",
    "    f = S         # Assuming all inbreeding is self fertilization \n",
    "    T = 1-S\n",
    "    \n",
    "    # Hardy-Weinberg genotype frequencies\n",
    "    P0 = p**2\n",
    "    Q0 = q**2\n",
    "    H0 = 2*p*q\n",
    "    \n",
    "    #Creating lists for parameters + H-W check\n",
    "    p_list, q_list, P1_list, Q1_list, H1_list, hwcheck = [],[],[],[],[],[]\n",
    "    p_list.append(p)\n",
    "    q_list.append(q)\n",
    "    P1_list.append(P0)\n",
    "    Q1_list.append(Q0)\n",
    "    H1_list.append(H0) \n",
    "    # Natural selection parameters with regard to self fertilization\n",
    "    w_11     = 1.0\n",
    "    w_12     = 1-(h*s)\n",
    "    w_22     = 1-s\n",
    "    w_bar    = ((p**2*w_11)+(2*p*q*w_12)+(q**2*w_22))\n",
    "    w_bar_f  = w_bar + f*p*q*(w_11+w_22-2*w_12)\n",
    "    \n",
    "    # Accounting for self fertilization \n",
    "    P1 = ((T*p**2+S*(P0+(0.25*H0)))*w_11)/w_bar_f\n",
    "    Q1 = ((T*q**2+S*(Q0+(0.25*H0)))*w_22)/w_bar_f\n",
    "    H1 = ((2*T*p*q+0.5*S*H0)*w_12)/w_bar_f\n",
    "    P1_list.append(P1)\n",
    "    Q1_list.append(Q1)\n",
    "    H1_list.append(H1) \n",
    "\n",
    "    \n",
    "    # Loop over generations, pick random gametes each generation, \n",
    "    # calculate new frequencies. \n",
    "    for generation in generations:\n",
    "         p_copies = 0\n",
    "         q_copies = 0\n",
    "         for gamete in range(2*N):\n",
    "             if random.random() < p:\n",
    "                 p_copies += 1\n",
    "             else:\n",
    "                 q_copies += 1 \n",
    "                \n",
    "        # Creating new allele frequencies and saving them\n",
    "         p = p_copies/(2*N)\n",
    "         q = q_copies/(2*N)\n",
    "         p_list.append(p)\n",
    "         q_list.append(q)\n",
    "             \n",
    "        # Creating new genotype frequencies and saving them, \n",
    "         P1 = ((T*p**2+S*(P1+(0.25*H1)))*w_11)/w_bar_f\n",
    "         Q1 = ((T*q**2+S*(Q1+(0.25*H1)))*w_22)/w_bar_f\n",
    "         H1 = ((2*T*p*q+0.5*S*H1)*w_12)/w_bar_f\n",
    "         p  = P1 + 0.5*H1\n",
    "         q  = 1-p\n",
    "         checksum = P1+H1+Q1\n",
    "         P1 = P1/checksum\n",
    "         Q1 = Q1/checksum\n",
    "         H1 = H1/checksum\n",
    "         P1_list.append(P1)\n",
    "         Q1_list.append(Q1)\n",
    "         H1_list.append(H1)\n",
    "         hwcheck.append(P1+H1+Q1)\n",
    "        \n",
    "    return P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting multiple repeats\n",
    "fig, ax = plt.subplots(2,2, figsize=(10,7), constrained_layout=True)\n",
    "ax[1,1].remove()\n",
    "repeats = range(5)\n",
    "for repeat in repeats:\n",
    "    P1_list, Q1_list, H1_list, p_list, q_list, hwcheck, generations = rand_self_fertilization_simulation(\n",
    "        A1_frequency          = 0.1,\n",
    "        A2_dominance_degree   = 0.2,\n",
    "        selection_coefficient = 0.2,\n",
    "        self_fert_factor      = 0.5,\n",
    "        population_size       = 5,\n",
    "        no_of_generations     = 250)\n",
    "    \n",
    "    ax[0,0].plot(generations, P1_list[:-2], label = 'Repeat: '+str(repeat))\n",
    "    ax[0,1].plot(generations, Q1_list[:-2], label = 'Repeat: '+str(repeat))\n",
    "    ax[1,0].plot(generations, H1_list[:-2], label = 'Repeat: '+str(repeat))\n",
    "\n",
    "# P Plot\n",
    "ax[0,0].set_title('P Genotype Frequency', fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[0,0].set_xlabel('No. of generations',  fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,0].set_ylabel('Genotype Frequency',  fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,0].set_ylim([0,1.0])\n",
    "ax[0,0].set_xlim([0,generations[-1]])\n",
    "ax[0,0].grid(axis='y', linestyle= '--')\n",
    "ax[0,0].grid(axis='x', linestyle= '--')\n",
    "ax[0,0].set_xticks(generations[::50])\n",
    "ax[0,0].legend()\n",
    "\n",
    "# Q Plot\n",
    "ax[0,1].set_title('Q Genotype Frequency', fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[0,1].set_xlabel('No. of generations',  fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,1].set_ylabel('Genotype Frequency',  fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[0,1].set_ylim([0,1.0])\n",
    "ax[0,1].set_xlim([0,generations[-1]])\n",
    "ax[0,1].grid(axis='y', linestyle= '--')\n",
    "ax[0,1].grid(axis='x', linestyle= '--')\n",
    "ax[0,1].set_xticks(generations[::50])\n",
    "ax[0,1].legend()\n",
    "\n",
    "# H Plot\n",
    "ax[1,0].set_title('H Genotype Frequency', fontdict = {'fontsize': 15, 'fontweight' : 600})\n",
    "ax[1,0].set_xlabel('No. of generations',  fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,0].set_ylabel('Genotype Frequency',  fontdict = {'fontsize': 10, 'fontweight' : 400})\n",
    "ax[1,0].set_ylim([0,1.0])\n",
    "ax[1,0].set_xlim([0,generations[-1]])\n",
    "ax[1,0].grid(axis='y', linestyle= '--')\n",
    "ax[1,0].grid(axis='x', linestyle= '--')\n",
    "ax[1,0].set_xticks(generations[::50])\n",
    "ax[1,0].legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
